packages/discord.js/src/client/websocket/WebSocketShard.js

```diff
@@ -837,25 +840,19 @@ class WebSocketShard extends EventEmitter {
         }
 
         // Emit the destroyed event if needed
-        if (emit) {
-          this._emitDestroyed();
-        } else if (
-          this.connection.readyState === WebSocket.CLOSING ||
-          this.connection.readyState === WebSocket.CLOSED
-        ) {
-          this.closeEmitted = false;
-          this.debug(
-            `[WebSocket] Adding a WebSocket close timeout to ensure a correct WS reconnect.
-            Timeout: ${this.manager.client.options.closeTimeout}ms`,
-          );
-          this.setWsCloseTimeout(this.manager.client.options.closeTimeout);
-        }
+        if (emit) this._emitDestroyed();
       }
     } else if (emit) {
       // We requested a destroy, but we had no connection. Emit destroyed
       this._emitDestroyed();
     }
 
+    this.debug(
+      `[WebSocket] Adding a WebSocket close timeout to ensure a correct WS reconnect.
+        Timeout: ${this.manager.client.options.closeTimeout}ms`,
+    );
+    this.setWsCloseTimeout(this.manager.client.options.closeTimeout);
```

Comment:
Moving this out of the `if (this.connection.readyState === WebSocket.OPEN) {}` block reintroduces the issue where the timeout would wait for closes on websockets that already were closed before and try to reconnect after we already successfully reconnected.
Either this should go back inside the if or at least the fix of my PR you supersede here to clear the timeout in the connect method before connecting (if one still exists) should be added in this PR too.

Reply:
> the timeout would wait for closes on WebSockets that already were closed before
I have tested this case already, `this.closeEmitted` will clear the timeout if the close was ever emitted before.

Reply:
<details>
<summary>
Logs
</summary>

```js
[ DJS ]: | Sun, 01 Jan 2023 12:57:20 | debug | [WS => Shard 25] | [RECONNECT] Discord asked us to reconnect
[ DJS ]: | Sun, 01 Jan 2023 12:57:20 | debug | [WS => Shard 25] | [DESTROY]
    Close Code    : 4000
    Reset         : false
    Emit DESTROYED: true
[ DJS ]: | Sun, 01 Jan 2023 12:57:20 | debug | [WS => Shard 25] | Clearing the heartbeat interval.
[ DJS ]: | Sun, 01 Jan 2023 12:57:20 | debug | [WS => Shard 25] | [WebSocket] Destroy: Attempting to close the WebSocket. | WS State: OPEN
[ DJS ]: | Sun, 01 Jan 2023 12:57:20 | debug | [WS => Shard 25] | [WebSocket] Close: Tried closing. | WS State: CLOSING
[ DJS ]: | Sun, 01 Jan 2023 12:57:20 | debug | [WS => Shard 25] | [WebSocket] Adding a WebSocket close timeout to ensure a correct WS reconnect.
        Timeout: 2500ms
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | [WebSocket] Clearing the close timeout.
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | [CLOSE]
    Event Code: 4000
    Clean     : true
    Reason    : 
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | Session id is present, attempting an immediate reconnect...
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | [CONNECT]
    Gateway    : wss://gateway.discord.gg/
    Version    : 9
    Encoding   : json
    Compression: none
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | Setting a HELLO timeout for 20s.
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | [CONNECTED] Took 198ms
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | Clearing the HELLO timeout.
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | Setting a heartbeat interval for 41250ms.
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | [RESUME] Session ~~, sequence 1384607
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | [RESUMED] Session ~~ | Replayed 1 events.
```
</details>

This happened 804 times in the past 3 days 13 hrs.

Also, the issue you were talking about is this #8984 from @itzJOHv and this should be fixed with this pr.

Reply:
For **testing**, you can also try to create a 4009 (zombie connection) yourself by not acknowledging the heartbeat from our side. Or the other way around, don't send the heartbeat which will cause a zombie connection again.
```js
  ackHeartbeat() {
    this.lastHeartbeatAcked = false;
```
<details>
<summary>
Logs
</summary>

```
//  node .\packages\discord.js\test\shard.js
kek dying
[WS => Shard 0] [HeartbeatTimer] Didn't receive a heartbeat ack last time, assuming zombie connection. Destroying and reconnecting.
    Status          : 0
    Sequence        : 2
    Connection State: OPEN
[WS => Shard 0] [DESTROY]
    Close Code    : 4009
    Reset         : true
    Emit DESTROYED: true
[WS => Shard 0] Clearing the heartbeat interval.
[WS => Shard 0] [WebSocket] Destroy: Attempting to close the WebSocket. | WS State: OPEN
[WS => Shard 0] [WebSocket] Close: Tried closing. | WS State: CLOSING
[WS => Shard 0] [WebSocket] Adding a WebSocket close timeout to ensure a correct WS reconnect.
        Timeout: 5000ms
[WS => Shard 0] [WebSocket] Clearing the close timeout.
[WS => Shard 0] [CLOSE]
    Event Code: 4009
    Clean     : true
    Reason    :
[WS => Shard 0] [CONNECT]
    Gateway    : wss://gateway.discord.gg/
    Version    : 10
    Encoding   : json
    Compression: zlib-stream
[WS => Shard 0] Setting a HELLO timeout for 20s.
[WS => Shard 0] [CONNECTED] Took 353ms
[WS => Shard 0] Clearing the HELLO timeout.
[WS => Shard 0] Setting a heartbeat interval for 41250ms.
[WS => Shard 0] [IDENTIFY] Shard 0/1 with intents: 513
[WS => Shard 0] [READY] Session a90b802c577369648e9a96d97c54994f | Resume url wss://gateway-us-east1-d.discord.gg.
[WS => Shard 0] [ReadyHeartbeat] Sending a heartbeat.
[WS => Shard 0] Shard received all its guilds. Marking as fully ready.
[WS => Shard 0] Heartbeat acknowledged, latency of 303ms.
```
</details>
And doing a zombie connection from our side did a successful reconnect.

Before I had set `this.closeEmitted` as false just before initiating a WsCloseTimeout which caused the issue that you are trying to fix in #8981
But now `this.closeEmitted` is set only twice, once in the onClose `this.closeEmitted = true` and once in the WsCloseTimeout `this.closeEmitted = false`

If you are still unsure whether this pr is going to cause issues please provide me the tests/logs in which it does

Reply:
One can create these kind of conditions by editing the Ws code accordingly.

Edge cases that I tested with this pr:
- not ack from our side
- not being ack from the discord side (not sending the ping to discord)
- close never being invoked
- all 3 above combined

Reply:
> websocket connection being already closed
Won't the `closeEmitted` be true if the websocket already was closed? https://github.com/discordjs/discord.js/blob/a7fd2d8194677d6e90cda96194b29ace9f8238af/packages/discord.js/src/client/websocket/WebSocketShard.js#L595

This will clear the timeout and return early.

If you can provide me a minimal repro using this pr so I can look into it.

Reply:
Also, the logs that you gave have error code 1011 which is not from the current code(pr). It will be helpful if you can get the exact same results with this current pr.

Reply:
We should be fine sorting this out on a follow-up if anything comes up.